File: G_Projects/G_Inventory_Management_System/ProductFactory/ProductFactory.java
package G_Projects.G_Inventory_Management_System.ProductFactory;

import G_Projects.G_Inventory_Management_System.CommonEnums.ProductEnum;
import G_Projects.G_Inventory_Management_System.ProductFactory.ConcreteProducts.ClothingProduct;
import G_Projects.G_Inventory_Management_System.ProductFactory.ConcreteProducts.ElectronicsProduct;
import G_Projects.G_Inventory_Management_System.ProductFactory.ConcreteProducts.GroceryProduct;

public class ProductFactory {
    public Product createProduct(ProductEnum category, String id, String name, double price, int quantity, int threshold){
        switch(category){
            case ELECTRONICS:
                return new ElectronicsProduct(id, name, price, quantity, threshold);
            case CLOTHING:
                return new ClothingProduct(id, name, price, quantity, threshold);
            case GROCERY:
                return new GroceryProduct(id, name, price, quantity, threshold);
            default:
                throw new IllegalArgumentException("Unsupported product category: " + category);
        }
    }
}

File: Python/A_Project/D_Inventory_Management_System/ProductFactory/product_factory.py
from CommonEnums.product_enum import ProductEnum
from ProductFactory.ConcreteProducts.clothing_product import ClothingProduct
from ProductFactory.ConcreteProducts.electronics_product import ElectronicsProduct
from ProductFactory.ConcreteProducts.grocery_product import GroceryProduct
from ProductFactory.product import Product


class ProductFactory:
    def create_product(self, category, product_id, name, price, quantity, threshold):
        if category == ProductEnum.ELECTRONICS:
            return ElectronicsProduct(product_id, name, price, quantity, threshold)
        if category == ProductEnum.CLOTHING:
            return ClothingProduct(product_id, name, price, quantity, threshold)
        if category == ProductEnum.GROCERY:
            return GroceryProduct(product_id, name, price, quantity, threshold)
        if category in (ProductEnum.FURNITURE, ProductEnum.OTHER):
            product = Product()
            product.set_id(product_id)
            product.set_name(name)
            product.set_price(price)
            product.set_quantity(quantity)
            product.set_threshold(threshold)
            product.set_category(category)
            return product
        raise ValueError(f"Unsupported product category: {category}")
